### ISSUE TITLE
`float` implicitly broadcasts

### ISSUE THREAD

**Issue by oscardssmith** [JuliaLang/julia#49929]
`float([2,3]) == [2.0, 3.0]`, but we should probably require this to be spelled `float.([2,3])` (and my guess is this deprecation was missed in the 0.7 cycle. In the meantime, should we deprecate this?

**Comment by KristofferC** [comment]
This is intentional. `float` here is not broadcasted but takes a vector as the domain of its input.

**Comment by oscardssmith** [comment]
right, but to me this really feels like broadcasting. It's not like it's converting a matrix to a float. It's taking all the entries and converting them separately.

**Comment by KristofferC** [comment]
This has been discussed before. I can't find the issue now but I'm pretty sure @stevengj was involved.

**Comment by adienes** [comment]
compare to `big([1,2,3])` and `string([1,2,3])`

**Comment by adienes** [comment]
even given those reasons to keep the functionality, it should still be `float!` and `complex!` though right? for the same reason as in https://github.com/JuliaLang/julia/issues/32528 that `String(::Vector{UInt8})` should be named `String!`

**Comment by KristofferC** [comment]
It doesn't mutate the input argument so why should it have a `!`?

**Comment by giordano** [comment]
> given that float(x) seems now to do the same thing as float.(x) 

It doesn't the same thing, float(array) is no-op when array is already all floats, broadcast always creates a new array (and I've taken advantage of this in some packages).

**Comment by adienes** [comment]
to me that seems like the worst and most dangerous of all worlds. it sometimes copies and sometimes doesn't?

I guess so
```
julia> x = rand(3);

julia> float(x)[1] = 100;

julia> x
3-element Vector{Float64}:
 100.0
   0.3173450352923628
   0.5859919943733706

julia> 

julia> x = rand(1:10, 3);

julia> float(x)[1] = 100;

julia> x
3-element Vector{Int64}:
 2
 3
 9
```

**Comment by timholy** [comment]
It's the same thing than `convert(AbstractVector{Float64}, v)` would do:
```julia
julia> v = [1.0,2.0,3.0]
3-element Vector{Float64}:
 1.0
 2.0
 3.0

julia> convert(AbstractVector{Float64}, v) === v
true
```

**Comment by adienes** [comment]
I did a little poking around and comparing to similar functions and I apologize for + retract some of my initial incredulity. here's what I got:

| Function(s) | aliasing | explicitly documented as such |
| -------- | -------- | -------- |
| `convert(typeof(x), x)` | usually | yes |
| `collect(eltype(x), x)` and `similar(eltype(t), t)` | never | no |
| `real(x)` and `conj(x)` | always when `<: Real`, otherwise never | yes |
| `complex(x)` | yes for complex-y type, otherwise usually not | no |
| `imag(x)` | never | yes |
| `float(x)` |  yes for float-y types, otherwise usually not| no |


Actionable:
* document `complex` and `float` the same way as `real` and `conj`, done in this PR: https://github.com/JuliaLang/julia/pull/49949

As for this issue, while it seems a bit more reasonable now, I still think the argument for `float` as a method is sort of weak. All of the related methods `real`, `imag`, `conj`, and `complex` work with integral types as well as float types, so `float` is kind of the only odd one out that is about precision more so than number domain. That is, by the same argument to keep `float` as a copy-free convert, I think one could argue that e.g. `trunc(Int, AbstractArray{<:Integer})` should also alias (which it does not).
